第一章 概述/lidar_localization/third_part/GeographicLib-1.52/man/GeodSolve.1
.\" Automatically generated by Pod::Man 4.14 (Pod::Simple 3.42)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GEODSOLVE 1"
.TH GEODSOLVE 1 "2021-06-21" "GeographicLib 1.52" "GeographicLib Utilities"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
GeodSolve \-\- perform geodesic calculations
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBGeodSolve\fR
[ \fB\-i\fR | \fB\-L\fR \fIlat1\fR \fIlon1\fR \fIazi1\fR |
\&\fB\-D\fR \fIlat1\fR \fIlon1\fR \fIazi1\fR \fIs13\fR | \fB\-I\fR \fIlat1\fR \fIlon1\fR \fIlat3\fR \fIlon3\fR ]
[ \fB\-a\fR ] [ \fB\-e\fR \fIa\fR \fIf\fR ] [ \fB\-u\fR ] [ \fB\-F\fR ]
[ \fB\-d\fR | \fB\-:\fR ] [ \fB\-w\fR ] [ \fB\-b\fR ] [ \fB\-f\fR ] [ \fB\-p\fR \fIprec\fR ] [ \fB\-E\fR ]
[ \fB\-\-comment\-delimiter\fR \fIcommentdelim\fR ]
[ \fB\-\-version\fR | \fB\-h\fR | \fB\-\-help\fR ]
[ \fB\-\-input\-file\fR \fIinfile\fR | \fB\-\-input\-string\fR \fIinstring\fR ]
[ \fB\-\-line\-separator\fR \fIlinesep\fR ]
[ \fB\-\-output\-file\fR \fIoutfile\fR ]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The shortest path between two points on the ellipsoid at (\fIlat1\fR,
\&\fIlon1\fR) and (\fIlat2\fR, \fIlon2\fR) is called the geodesic.  Its length is
\&\fIs12\fR and the geodesic from point 1 to point 2 has forward azimuths
\&\fIazi1\fR and \fIazi2\fR at the two end points.
.PP
\&\fBGeodSolve\fR operates in one of three modes:
.IP "1." 4
By default, \fBGeodSolve\fR accepts lines on the standard input containing
\&\fIlat1\fR \fIlon1\fR \fIazi1\fR \fIs12\fR and prints \fIlat2\fR \fIlon2\fR \fIazi2\fR
on standard output.  This is the direct geodesic calculation.
.IP "2." 4
With the \fB\-i\fR command line argument, \fBGeodSolve\fR performs the inverse
geodesic calculation.  It reads lines containing \fIlat1\fR \fIlon1\fR \fIlat2\fR
\&\fIlon2\fR and prints the corresponding values of \fIazi1\fR \fIazi2\fR \fIs12\fR.
.IP "3." 4
Command line arguments \fB\-L\fR \fIlat1\fR \fIlon1\fR \fIazi1\fR specify a geodesic
line.  \fBGeodSolve\fR then accepts a sequence of \fIs12\fR values (one per
line) on standard input and prints \fIlat2\fR \fIlon2\fR \fIazi2\fR for each.
This generates a sequence of points on a single geodesic.  Command line
arguments \fB\-D\fR and \fB\-I\fR work similarly with the geodesic line defined
in terms of a direct or inverse geodesic calculation, respectively.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-i\fR" 4
.IX Item "-i"
perform an inverse geodesic calculation (see 2 above).
.IP "\fB\-L\fR \fIlat1\fR \fIlon1\fR \fIazi1\fR" 4
.IX Item "-L lat1 lon1 azi1"
line mode (see 3 above); generate a sequence of points along the
geodesic specified by \fIlat1\fR \fIlon1\fR \fIazi1\fR.  The \fB\-w\fR flag can be
used to swap the default order of the 2 geographic coordinates, provided
that it appears before \fB\-L\fR.  (\fB\-l\fR is an alternative, deprecated,
spelling of this flag.)
.IP "\fB\-D\fR \fIlat1\fR \fIlon1\fR \fIazi1\fR \fIs13\fR" 4
.IX Item "-D lat1 lon1 azi1 s13"
line mode (see 3 above); generate a sequence of points along the
geodesic specified by \fIlat1\fR \fIlon1\fR \fIazi1\fR \fIs13\fR.  The \fB\-w\fR flag
can be used to swap the default order of the 2 geographic coordinates,
provided that it appears before \fB\-D\fR.  Similarly, the \fB\-a\fR flag can be
used to change the interpretation of \fIs13\fR to \fIa13\fR, provided that it
appears before \fB\-D\fR.
.IP "\fB\-I\fR \fIlat1\fR \fIlon1\fR \fIlat3\fR \fIlon3\fR" 4
.IX Item "-I lat1 lon1 lat3 lon3"
line mode (see 3 above); generate a sequence of points along the
geodesic specified by \fIlat1\fR \fIlon1\fR \fIlat3\fR \fIlon3\fR.  The \fB\-w\fR flag
can be used to swap the default order of the 2 geographic coordinates,
provided that it appears before \fB\-I\fR.
.IP "\fB\-a\fR" 4
.IX Item "-a"
toggle the arc mode flag (it starts off); if this flag is on, then on
input \fIand\fR output \fIs12\fR is replaced by \fIa12\fR the arc length (in
degrees) on the auxiliary sphere.  See \*(L"\s-1AUXILIARY SPHERE\*(R"\s0.
.IP "\fB\-e\fR \fIa\fR \fIf\fR" 4
.IX Item "-e a f"
specify the ellipsoid via the equatorial radius, \fIa\fR and
the flattening, \fIf\fR.  Setting \fIf\fR = 0 results in a sphere.  Specify
\&\fIf\fR < 0 for a prolate ellipsoid.  A simple fraction, e.g., 1/297,
is allowed for \fIf\fR.  By default, the \s-1WGS84\s0 ellipsoid is used, \fIa\fR =
6378137 m, \fIf\fR = 1/298.257223563.
.IP "\fB\-u\fR" 4
.IX Item "-u"
unroll the longitude.  Normally, on output longitudes are reduced to lie
in [\-180deg,180deg).  However with this option, the returned longitude
\&\fIlon2\fR is \*(L"unrolled\*(R" so that \fIlon2\fR \- \fIlon1\fR indicates how often and
in what sense the geodesic has encircled the earth.  Use the \fB\-f\fR
option, to get both longitudes printed.
.IP "\fB\-F\fR" 4
.IX Item "-F"
fractional mode.  This only has any effect with the \fB\-D\fR and \fB\-I\fR
options (and is otherwise ignored).  The values read on standard input
are interpreted as fractional distances to point 3, i.e., as
\&\fIs12\fR/\fIs13\fR instead of \fIs12\fR.  If arc mode is in effect, then the
values denote fractional arc length, i.e., \fIa12\fR/\fIa13\fR.  The
fractional distances can be entered as a simple fraction, e.g., 3/4.
.IP "\fB\-d\fR" 4
.IX Item "-d"
output angles as degrees, minutes, seconds instead of decimal degrees.
.IP "\fB\-:\fR" 4
.IX Item "-:"
like \fB\-d\fR, except use : as a separator instead of the d, ', and "
delimiters.
.IP "\fB\-w\fR" 4
.IX Item "-w"
toggle the longitude first flag (it starts off); if the flag is on, then
on input and output, longitude precedes latitude (except that, on input,
this can be overridden by a hemisphere designator, \fIN\fR, \fIS\fR, \fIE\fR,
\&\fIW\fR).
.IP "\fB\-b\fR" 4
.IX Item "-b"
report the \fIback\fR azimuth at point 2 instead of the forward azimuth.
.IP "\fB\-f\fR" 4
.IX Item "-f"
full output; each line of output consists of 12 quantities: \fIlat1\fR
\&\fIlon1\fR \fIazi1\fR \fIlat2\fR \fIlon2\fR \fIazi2\fR \fIs12\fR \fIa12\fR \fIm12\fR \fIM12\fR
\&\fIM21\fR \fIS12\fR.  \fIa12\fR is described in \*(L"\s-1AUXILIARY SPHERE\*(R"\s0.  The four
quantities \fIm12\fR, \fIM12\fR, \fIM21\fR, and \fIS12\fR are described in
\&\*(L"\s-1ADDITIONAL QUANTITIES\*(R"\s0.
.IP "\fB\-p\fR \fIprec\fR" 4
.IX Item "-p prec"
set the output precision to \fIprec\fR (default 3); \fIprec\fR is the
precision relative to 1 m.  See \*(L"\s-1PRECISION\*(R"\s0.
.IP "\fB\-E\fR" 4
.IX Item "-E"
use \*(L"exact\*(R" algorithms (based on elliptic integrals) for the geodesic
calculations.  These are more accurate than the (default) series
expansions for |\fIf\fR| > 0.02.
.IP "\fB\-\-comment\-delimiter\fR \fIcommentdelim\fR" 4
.IX Item "--comment-delimiter commentdelim"
set the comment delimiter to \fIcommentdelim\fR (e.g., \*(L"#\*(R" or \*(L"//\*(R").  If
set, the input lines will be scanned for this delimiter and, if found,
the delimiter and the rest of the line will be removed prior to
processing and subsequently appended to the output line (separated by a
space).
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
print version and exit.
.IP "\fB\-h\fR" 4
.IX Item "-h"
print usage and exit.
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
print full documentation and exit.
.IP "\fB\-\-input\-file\fR \fIinfile\fR" 4
.IX Item "--input-file infile"
read input from the file \fIinfile\fR instead of from standard input; a file
name of \*(L"\-\*(R" stands for standard input.
.IP "\fB\-\-input\-string\fR \fIinstring\fR" 4
.IX Item "--input-string instring"
read input from the string \fIinstring\fR instead of from standard input.
All occurrences of the line separator character (default is a semicolon)
in \fIinstring\fR are converted to newlines before the reading begins.
.IP "\fB\-\-line\-separator\fR \fIlinesep\fR" 4
.IX Item "--line-separator linesep"
set the line separator character to \fIlinesep\fR.  By default this is a
semicolon.
.IP "\fB\-\-output\-file\fR \fIoutfile\fR" 4
.IX Item "--output-file outfile"
write output to the file \fIoutfile\fR instead of to standard output; a
file name of \*(L"\-\*(R" stands for standard output.
.SH "INPUT"
.IX Header "INPUT"
\&\fBGeodSolve\fR measures all angles in degrees and all lengths (\fIs12\fR) in
meters, and all areas (\fIS12\fR) in meters^2.  On input angles (latitude,
longitude, azimuth, arc length) can be as decimal degrees or degrees,
minutes, seconds.  For example, \f(CW\*(C`40d30\*(C'\fR, \f(CW\*(C`40d30\*(Aq\*(C'\fR, \f(CW\*(C`40:30\*(C'\fR, \f(CW\*(C`40.5d\*(C'\fR,
and \f(CW40.5\fR are all equivalent.  By default, latitude precedes longitude
for each point (the \fB\-w\fR flag switches this convention); however on
input either may be given first by appending (or prepending) \fIN\fR or
\&\fIS\fR to the latitude and \fIE\fR or \fIW\fR to the longitude.  Azimuths are
measured clockwise from north; however this may be overridden with \fIE\fR
or \fIW\fR.
.PP
For details on the allowed formats for angles, see the \f(CW\*(C`GEOGRAPHIC
COORDINATES\*(C'\fR section of \fBGeoConvert\fR\|(1).
.SH "AUXILIARY SPHERE"
.IX Header "AUXILIARY SPHERE"
Geodesics on the ellipsoid can be transferred to the \fIauxiliary sphere\fR
on which the distance is measured in terms of the arc length \fIa12\fR
(measured in degrees) instead of \fIs12\fR.  In terms of \fIa12\fR, 180
degrees is the distance from one equator crossing to the next or from
the minimum latitude to the maximum latitude.  Geodesics with \fIa12\fR
> 180 degrees do not correspond to shortest paths.  With the \fB\-a\fR
flag, \fIs12\fR (on both input and output) is replaced by \fIa12\fR.  The
\&\fB\-a\fR flag does \fInot\fR affect the full output given by the \fB\-f\fR flag
(which always includes both \fIs12\fR and \fIa12\fR).
.SH "ADDITIONAL QUANTITIES"
.IX Header "ADDITIONAL QUANTITIES"
The \fB\-f\fR flag reports four additional quantities.
.PP
The reduced length of the geodesic, \fIm12\fR, is defined such that if the
initial azimuth is perturbed by d\fIazi1\fR (radians) then the second point
is displaced by \fIm12\fR d\fIazi1\fR in the direction perpendicular to the
geodesic.  \fIm12\fR is given in meters.  On a curved surface the
reduced length obeys a symmetry relation, \fIm12\fR + \fIm21\fR = 0.  On a
flat surface, we have \fIm12\fR = \fIs12\fR.
.PP
\&\fIM12\fR and \fIM21\fR are geodesic scales.  If two geodesics are parallel at
point 1 and separated by a small distance \fIdt\fR, then they are separated

第一章 概述/lidar_localization/third_part/GeographicLib-1.52/tools/GeodSolve.cpp
/**
 * \file GeodSolve.cpp
 * \brief Command line utility for geodesic calculations
 *
 * Copyright (c) Charles Karney (2009-2019) <charles@karney.com> and licensed
 * under the MIT/X11 License.  For more information, see
 * https://geographiclib.sourceforge.io/
 *
 * See the <a href="GeodSolve.1.html">man page</a> for usage information.
 **********************************************************************/

#include <iostream>
#include <string>
#include <sstream>
#include <fstream>
#include <GeographicLib/Geodesic.hpp>
#include <GeographicLib/GeodesicLine.hpp>
#include <GeographicLib/GeodesicExact.hpp>
#include <GeographicLib/GeodesicLineExact.hpp>
#include <GeographicLib/DMS.hpp>
#include <GeographicLib/Utility.hpp>

#if defined(_MSC_VER)
// Squelch warnings about constant conditional expressions and potentially
// uninitialized local variables
#  pragma warning (disable: 4127 4701)
#endif

#include "GeodSolve.usage"

typedef GeographicLib::Math::real real;

std::string LatLonString(real lat, real lon, int prec, bool dms, char dmssep,
                         bool longfirst) {
  using namespace GeographicLib;
  std::string
    latstr = dms ? DMS::Encode(lat, prec + 5, DMS::LATITUDE, dmssep) :
    DMS::Encode(lat, prec + 5, DMS::NUMBER),
    lonstr = dms ? DMS::Encode(lon, prec + 5, DMS::LONGITUDE, dmssep) :
    DMS::Encode(lon, prec + 5, DMS::NUMBER);
  return
    (longfirst ? lonstr : latstr) + " " + (longfirst ? latstr : lonstr);
}

std::string AzimuthString(real azi, int prec, bool dms, char dmssep) {
  using namespace GeographicLib;
  return dms ? DMS::Encode(azi, prec + 5, DMS::AZIMUTH, dmssep) :
    DMS::Encode(azi, prec + 5, DMS::NUMBER);
}

std::string DistanceStrings(real s12, real a12,
                            bool full, bool arcmode, int prec, bool dms) {
  using namespace GeographicLib;
  std::string s;
  if (full || !arcmode)
    s += Utility::str(s12, prec);
  if (full)
    s += " ";
  if (full || arcmode)
    s += DMS::Encode(a12, prec + 5, dms ? DMS::NONE : DMS::NUMBER);
  return s;
}

real ReadDistance(const std::string& s, bool arcmode, bool fraction = false) {
  using namespace GeographicLib;
  return fraction ? Utility::fract<real>(s) :
    (arcmode ? DMS::DecodeAngle(s) : Utility::val<real>(s));
}

int main(int argc, const char* const argv[]) {
  try {
    using namespace GeographicLib;
    enum { NONE = 0, LINE, DIRECT, INVERSE };
    Utility::set_digits();
    bool inverse = false, arcmode = false,
      dms = false, full = false, exact = false, unroll = false,
      longfirst = false, azi2back = false, fraction = false,
      arcmodeline = false;
    real
      a = Constants::WGS84_a(),
      f = Constants::WGS84_f();
    real lat1, lon1, azi1, lat2, lon2, azi2, s12, m12, a12, M12, M21, S12,
      mult = 1;
    int linecalc = NONE, prec = 3;
    std::string istring, ifile, ofile, cdelim;
    char lsep = ';', dmssep = char(0);

    for (int m = 1; m < argc; ++m) {
      std::string arg(argv[m]);
      if (arg == "-i") {
        inverse = true;
        linecalc = NONE;
      } else if (arg == "-a")
        arcmode = !arcmode;
      else if (arg == "-F")
        fraction = true;
      else if (arg == "-L" || arg == "-l") { // -l is DEPRECATED
        inverse = false;
        linecalc = LINE;
        if (m + 3 >= argc) return usage(1, true);
        try {
          DMS::DecodeLatLon(std::string(argv[m + 1]), std::string(argv[m + 2]),
                            lat1, lon1, longfirst);
          azi1 = DMS::DecodeAzimuth(std::string(argv[m + 3]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -L: " << e.what() << "\n";
          return 1;
        }
        m += 3;
      } else if (arg == "-D") {
        inverse = false;
        linecalc = DIRECT;
        if (m + 4 >= argc) return usage(1, true);
        try {
          DMS::DecodeLatLon(std::string(argv[m + 1]), std::string(argv[m + 2]),
                            lat1, lon1, longfirst);
          azi1 = DMS::DecodeAzimuth(std::string(argv[m + 3]));
          s12 = ReadDistance(std::string(argv[m + 4]), arcmode);
          arcmodeline = arcmode;
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -D: " << e.what() << "\n";
          return 1;
        }
        m += 4;
      } else if (arg == "-I") {
        inverse = false;
        linecalc = INVERSE;
        if (m + 4 >= argc) return usage(1, true);
        try {
          DMS::DecodeLatLon(std::string(argv[m + 1]), std::string(argv[m + 2]),
                            lat1, lon1, longfirst);
          DMS::DecodeLatLon(std::string(argv[m + 3]), std::string(argv[m + 4]),
                            lat2, lon2, longfirst);
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -I: " << e.what() << "\n";
          return 1;
        }
        m += 4;
      } else if (arg == "-e") {
        if (m + 2 >= argc) return usage(1, true);
        try {
          a = Utility::val<real>(std::string(argv[m + 1]));
          f = Utility::fract<real>(std::string(argv[m + 2]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -e: " << e.what() << "\n";
          return 1;
        }
        m += 2;
      } else if (arg == "-u")
        unroll = true;
      else if (arg == "-d") {
        dms = true;
        dmssep = '\0';
      } else if (arg == "-:") {
        dms = true;
        dmssep = ':';
      } else if (arg == "-w")
        longfirst = !longfirst;
      else if (arg == "-b")
        azi2back = true;
      else if (arg == "-f")
        full = true;
      else if (arg == "-p") {
        if (++m == argc) return usage(1, true);
        try {
          prec = Utility::val<int>(std::string(argv[m]));
        }
        catch (const std::exception&) {
          std::cerr << "Precision " << argv[m] << " is not a number\n";
          return 1;
        }
      } else if (arg == "-E")
        exact = true;
      else if (arg == "--input-string") {
        if (++m == argc) return usage(1, true);
        istring = argv[m];
      } else if (arg == "--input-file") {
        if (++m == argc) return usage(1, true);
        ifile = argv[m];
      } else if (arg == "--output-file") {
        if (++m == argc) return usage(1, true);
        ofile = argv[m];
      } else if (arg == "--line-separator") {
        if (++m == argc) return usage(1, true);
        if (std::string(argv[m]).size() != 1) {
          std::cerr << "Line separator must be a single character\n";
          return 1;
        }
        lsep = argv[m][0];
      } else if (arg == "--comment-delimiter") {
        if (++m == argc) return usage(1, true);
        cdelim = argv[m];
      } else if (arg == "--version") {
        std::cout << argv[0] << ": GeographicLib version "
                  << GEOGRAPHICLIB_VERSION_STRING << "\n";
        return 0;
      } else
        return usage(!(arg == "-h" || arg == "--help"), arg != "--help");
    }

    if (!ifile.empty() && !istring.empty()) {
      std::cerr << "Cannot specify --input-string and --input-file together\n";
      return 1;
    }
    if (ifile == "-") ifile.clear();
    std::ifstream infile;
    std::istringstream instring;
    if (!ifile.empty()) {
      infile.open(ifile.c_str());
      if (!infile.is_open()) {
        std::cerr << "Cannot open " << ifile << " for reading\n";
        return 1;
      }
    } else if (!istring.empty()) {
      std::string::size_type m = 0;
      while (true) {
        m = istring.find(lsep, m);
        if (m == std::string::npos)
          break;
        istring[m] = '\n';
      }
      instring.str(istring);
    }
    std::istream* input = !ifile.empty() ? &infile :
      (!istring.empty() ? &instring : &std::cin);

    std::ofstream outfile;
    if (ofile == "-") ofile.clear();
    if (!ofile.empty()) {
      outfile.open(ofile.c_str());
      if (!outfile.is_open()) {
        std::cerr << "Cannot open " << ofile << " for writing\n";
        return 1;
      }
    }
    std::ostream* output = !ofile.empty() ? &outfile : &std::cout;

    // GeodesicExact mask values are the same as Geodesic
    unsigned outmask = Geodesic::LATITUDE | Geodesic::LONGITUDE |
      Geodesic::AZIMUTH;        // basic output quantities
    outmask |= inverse ? Geodesic::DISTANCE : // distance-related flags
      (arcmode ? Geodesic::NONE : Geodesic::DISTANCE_IN);
    // longitude unrolling
    outmask |= unroll ? Geodesic::LONG_UNROLL : Geodesic::NONE;
    // full output -- don't use Geodesic::ALL since this includes DISTANCE_IN
    outmask |= full ? (Geodesic::DISTANCE | Geodesic::REDUCEDLENGTH |
                       Geodesic::GEODESICSCALE | Geodesic::AREA) :
      Geodesic::NONE;

    const Geodesic      geods(a, f);
    const GeodesicExact geode(a, f);
    GeodesicLine      ls;
    GeodesicLineExact le;
    if (linecalc) {
      if (linecalc == LINE) fraction = false;
      if (exact) {
        le = linecalc == DIRECT ?
          geode.GenDirectLine(lat1, lon1, azi1, arcmodeline, s12, outmask) :
          linecalc == INVERSE ?
          geode.InverseLine(lat1, lon1, lat2, lon2, outmask) :
          // linecalc == LINE
          geode.Line(lat1, lon1, azi1, outmask);
        mult = fraction ? le.GenDistance(arcmode) : 1;
        if (linecalc == INVERSE) azi1 = le.Azimuth();
      } else {
        ls = linecalc == DIRECT ?
          geods.GenDirectLine(lat1, lon1, azi1, arcmodeline, s12, outmask) :
          linecalc == INVERSE ?
          geods.InverseLine(lat1, lon1, lat2, lon2, outmask) :
          // linecalc == LINE
          geods.Line(lat1, lon1, azi1, outmask);
        mult = fraction ? ls.GenDistance(arcmode) : 1;
        if (linecalc == INVERSE) azi1 = ls.Azimuth();
      }
    }

    // Max precision = 10: 0.1 nm in distance, 10^-15 deg (= 0.11 nm),
    // 10^-11 sec (= 0.3 nm).
    prec = std::min(10 + Math::extra_digits(), std::max(0, prec));
    std::string s, eol, slat1, slon1, slat2, slon2, sazi1, ss12, strc;
    std::istringstream str;
    int retval = 0;
    while (std::getline(*input, s)) {
      try {
        eol = "\n";
        if (!cdelim.empty()) {
          std::string::size_type m = s.find(cdelim);
          if (m != std::string::npos) {
            eol = " " + s.substr(m) + "\n";
            s = s.substr(0, m);
          }
        }
        str.clear(); str.str(s);
        if (inverse) {
          if (!(str >> slat1 >> slon1 >> slat2 >> slon2))
            throw GeographicErr("Incomplete input: " + s);
          if (str >> strc)
            throw GeographicErr("Extraneous input: " + strc);
          DMS::DecodeLatLon(slat1, slon1, lat1, lon1, longfirst);
          DMS::DecodeLatLon(slat2, slon2, lat2, lon2, longfirst);
          a12 = exact ?
            geode.GenInverse(lat1, lon1, lat2, lon2, outmask,
                             s12, azi1, azi2, m12, M12, M21, S12) :
            geods.GenInverse(lat1, lon1, lat2, lon2, outmask,
                             s12, azi1, azi2, m12, M12, M21, S12);
          if (full) {
            if (unroll) {
              real e;
              lon2 = lon1 + Math::AngDiff(lon1, lon2, e);
              lon2 += e;
            } else {
              lon1 = Math::AngNormalize(lon1);
              lon2 = Math::AngNormalize(lon2);
            }
            *output << LatLonString(lat1, lon1, prec, dms, dmssep, longfirst)
                    << " ";
          }
          *output << AzimuthString(azi1, prec, dms, dmssep) << " ";
          if (full)
            *output << LatLonString(lat2, lon2, prec, dms, dmssep, longfirst)
                    << " ";
          if (azi2back)
            azi2 += azi2 >= 0 ? -180 : 180;
          *output << AzimuthString(azi2, prec, dms, dmssep) << " "
                  << DistanceStrings(s12, a12, full, arcmode, prec, dms);
          if (full)
            *output << " " << Utility::str(m12, prec)
                    << " " << Utility::str(M12, prec+7)
                    << " " << Utility::str(M21, prec+7)
                    << " " << Utility::str(S12, std::max(prec-7, 0));
          *output << eol;
        } else {
          if (linecalc) {
            if (!(str >> ss12))
              throw GeographicErr("Incomplete input: " + s);
            if (str >> strc)
              throw GeographicErr("Extraneous input: " + strc);
            // In fraction mode input is read as a distance
            s12 = ReadDistance(ss12, !fraction && arcmode, fraction) * mult;
            a12 = exact ?
              le.GenPosition(arcmode, s12, outmask,
                             lat2, lon2, azi2, s12, m12, M12, M21, S12) :
              ls.GenPosition(arcmode, s12, outmask,
                             lat2, lon2, azi2, s12, m12, M12, M21, S12);
          } else {
            if (!(str >> slat1 >> slon1 >> sazi1 >> ss12))
              throw GeographicErr("Incomplete input: " + s);
            if (str >> strc)
              throw GeographicErr("Extraneous input: " + strc);
            DMS::DecodeLatLon(slat1, slon1, lat1, lon1, longfirst);
            azi1 = DMS::DecodeAzimuth(sazi1);
            s12 = ReadDistance(ss12, arcmode);
            a12 = exact ?
              geode.GenDirect(lat1, lon1, azi1, arcmode, s12, outmask,
                              lat2, lon2, azi2, s12, m12, M12, M21, S12) :
              geods.GenDirect(lat1, lon1, azi1, arcmode, s12, outmask,
                              lat2, lon2, azi2, s12, m12, M12, M21, S12);
          }
          if (full)
            *output
              << LatLonString(lat1, unroll ? lon1 : Math::AngNormalize(lon1),
                              prec, dms, dmssep, longfirst)
              << " " << AzimuthString(azi1, prec, dms, dmssep) << " ";
          if (azi2back)
            azi2 += azi2 >= 0 ? -180 : 180;
          *output << LatLonString(lat2, lon2, prec, dms, dmssep, longfirst)
                  << " " << AzimuthString(azi2, prec, dms, dmssep);
          if (full)
            *output << " "
                    << DistanceStrings(s12, a12, full, arcmode, prec, dms)
                    << " " << Utility::str(m12, prec)
                    << " " << Utility::str(M12, prec+7)
                    << " " << Utility::str(M21, prec+7)
                    << " " << Utility::str(S12, std::max(prec-7, 0));
          *output << eol;
        }
      }
      catch (const std::exception& e) {
        // Write error message cout so output lines match input lines
        *output << "ERROR: " << e.what() << "\n";
        retval = 1;
      }
    }
    return retval;
  }
  catch (const std::exception& e) {
    std::cerr << "Caught exception: " << e.what() << "\n";
    return 1;
  }
  catch (...) {
    std::cerr << "Caught unknown exception\n";
    return 1;
  }
}
